Add WIP Media Session API updates article

beaufortfrancois opened the pull request:
This PR is still WIP and is used to get feedback from engineers.

WDYT @mounirlamouri @xxyzzzq?

xxyzzzq commented on src/content/en/updates/2017/02/media-session.md:
```diff
@@ -0,0 +1,374 @@
+    }
+    
+    function getNetworkArtwork(request) {
+       // Fetch network artwork.
```
nit: indent 2 spaces

petele replied:
No, this is fine as is.

mounirlamouri replied:
It looks like this is a 3-space indent and the other functions have a 2-space indent.

beaufortfrancois replied:
Fixed.

petele commented on src/content/en/updates/2017/02/media-session.md:
```diff
@@ -0,0 +1,374 @@
+
+{# wf_updated_on: 2017-02-06 #}
+{# wf_published_on: 2017-02-06 #}
+{# wf_tags: news,mediasession,play,pause #}
```
Build Errors for uncommon tags. You can add these tags in `src/data/commonTags.json`. Be sure to review the file first to make sure you're not creating similar tags.
* [ ] src/content/en/updates/2017/02/media-session.md Uncommon tag (mediasession) found.
* [ ] src/content/en/updates/2017/02/media-session.md Uncommon tag (pause) found.

petele replied:
Also add chrome57 to the list of tags.

beaufortfrancois replied:
I didn't know this file! Thank you ;)
Fixed.

beaufortfrancois replied:
Fixed.

petele commented on src/content/en/updates/2017/02/media-session.md:
```diff
@@ -0,0 +1,374 @@
+{# wf_updated_on: 2017-02-06 #}
+{# wf_published_on: 2017-02-06 #}
+{# wf_tags: news,mediasession,play,pause #}
+{# wf_featured_image: /web/updates/images/2017/02/featured.png #}
```
Build Error:
* [ ] src/content/en/updates/2017/02/media-session.md WF Tag `wf_featured_image` found, but couldn't find image - /web/updates/images/2017/02/featured.png

petele commented on src/content/en/updates/2017/02/media-session.md:
```diff
@@ -0,0 +1,374 @@
+{# wf_published_on: 2017-02-06 #}
+{# wf_tags: news,mediasession,play,pause #}
+{# wf_featured_image: /web/updates/images/2017/02/featured.png #}
+
```
I **strongly** encourage also including a `{# wf_featured_snippet: blah blah blah #}`.

The featured snipped is used on the index pages as the primary description. It CAN include HTML and is your "hook" to get someone to read your article. The description at the top of the document is used as the meta description and is only used as the primary description if `wf_featured_snippet` isn't provided.

beaufortfrancois replied:
Noted. Thanks!

petele commented on src/content/en/updates/2017/02/media-session.md:
```diff
@@ -0,0 +1,374 @@
+If you don't want to play audio right after the first interaction, I recommend
+you use the `load()` method of the audio element.
+
+<pre class="prettyprint">
```
Instead of wrapping in pretty print, the preferred method is to indent four spaces. I think this works, but there may also be some weirdness around it if I remember right.

beaufortfrancois replied:
In this particular code sample, I'm using `<strong>audio.load()</strong>` which is not parsed correctly when using four spaces indentation. If there's a way to get it without `prettyprint`, I'm all for it ;)

petele commented on src/content/en/updates/2017/02/media-session.md:
```diff
@@ -0,0 +1,374 @@
+<div class="clearfix"></div>
+<div class="attempt-left">
+  <figure>
+    <img src="https://placehold.it/350x350?text=With no Media Session" alt="TODO">
```
`alt="TODO"`

beaufortfrancois replied:
Fixed!

petele commented on src/content/en/updates/2017/02/media-session.md:
```diff
@@ -0,0 +1,374 @@
+</div>
+<div class="attempt-right">
+  <figure>
+    <img src="https://placehold.it/350x350?text=With Media Session" alt="TODO">
```
`alt="TODO"`

beaufortfrancois replied:
Fixed.

petele commented on src/content/en/updates/2017/02/media-session.md:
```diff
@@ -0,0 +1,374 @@
+<div class="clearfix"></div>
+<div class="attempt-left">
+  <figure>
+    <img src="https://placehold.it/350x350?text=Lock screen" alt="TODO">
```
`alt="TODO"`

beaufortfrancois replied:
Fixed.

petele commented on src/content/en/updates/2017/02/media-session.md:
```diff
@@ -0,0 +1,374 @@
+</div>
+<div class="attempt-right">
+  <figure>
+    <img src="https://placehold.it/350x350?text=Wear notification" alt="TODO">
```
`alt="TODO"`

beaufortfrancois replied:
Fixed

mounirlamouri commented on src/content/en/updates/2017/02/media-session.md:
```diff
@@ -0,0 +1,374 @@
+
+As you may know, `autoplay` is disabled on Chrome for Android which
+means we have to use the `play()` method of the audio element there. This
+method must be triggered by [a user gesture] such as a touch or a mouse click.
```
What does [ ] do here?

beaufortfrancois replied:
It automagically adds URLs when they're defined at the very bottom of the file.

mounirlamouri commented on src/content/en/updates/2017/02/media-session.md:
```diff
@@ -0,0 +1,374 @@
+events. In other words, the user must click on a button before your web app can
+actually make some noise.
+
+    playButton.addEventListener('pointerup', function(event) {
```
Why not 'click'?

mounirlamouri replied:
s/function (event)/e =>/ ?

beaufortfrancois replied:
`pointerup` is what we promote nowadays.

beaufortfrancois replied:
I tend to use arrow function for non `addEventListener` stuff as it creates confusion when it's not needed.

mounirlamouri commented on src/content/en/updates/2017/02/media-session.md:
```diff
@@ -0,0 +1,374 @@
+
+let audio = document.querySelector('audio');
+
+welcomeButton.addEventListener('pointerup', function(event) {
```
ditto

beaufortfrancois replied:
same answer.

mounirlamouri commented on src/content/en/updates/2017/02/media-session.md:
```diff
@@ -0,0 +1,374 @@
+    });
+
+If you don't want to play audio right after the first interaction, I recommend
+you use the `load()` method of the audio element.
```
You can explain that the reason is that Blink will keep track of whether the user interacted with the element in some ways and `load()` is one of them. It will also help the playback be more smooth because the content will be loaded.

beaufortfrancois replied:
Thanks! Explanation added.

mounirlamouri commented on src/content/en/updates/2017/02/media-session.md:
```diff
@@ -0,0 +1,374 @@
+Note: As the browser may consider the web app is not be playing when media
+files are seeking or loading, you can override this behaviour by setting
+`navigator.mediaSession.playbackState` to `"playing"` or `"paused"`.  This
+comes handy when you want to make sure your web app UI state is synced with the
```
typo: s/state/stay/

beaufortfrancois replied:
@mounirlamouri I don't see the typo ;(

beaufortfrancois replied:
Thanks!

mounirlamouri commented on src/content/en/updates/2017/02/media-session.md:
```diff
@@ -0,0 +1,374 @@
+resources. See the [Caching checklist].
+2. Make sure all media and artwork files are served with the
+`Allow-Control-Allow-Origin: *` HTTP header. This will allow third-party web
+apps to fetch and consume HTTP responses from your web server.
```
Do we want to advertise `*` or should we tell web developers that listing their website's origin is preferable?

beaufortfrancois replied:
That's a good question. @samdutton WDYT?
I'm favor of promoting `*` to help 3rd party web apps.

mounirlamouri commented on src/content/en/updates/2017/02/media-session.md:
```diff
@@ -0,0 +1,374 @@
+
+## Implementation nits
+
+- Chrome Media notifications show only if media files duration is [at least 5 seconds].
```
In case of, we don't take full audio focus if shorter than 5 seconds. We take the intermittent one.

beaufortfrancois replied:
@mounirlamouri I'm not sure to understand what you mean.

beaufortfrancois replied:
Got it.

mounirlamouri commented on src/content/en/updates/2017/02/media-session.md:
```diff
@@ -0,0 +1,374 @@
+## Implementation nits
+
+- Chrome Media notifications show only if media files duration is [at least 5 seconds].
+- Muted media files won't trigger Chrome media notifications.
```
That's not entirely true. `<video autoplay muted>` will not. Anything else muted will. Even the former case will if it gets unmuted then re-muted.

beaufortfrancois replied:
In that case, I'll remove this nit.

mounirlamouri commented on src/content/en/updates/2017/02/media-session.md:
```diff
@@ -0,0 +1,374 @@
+
+- Chrome Media notifications show only if media files duration is [at least 5 seconds].
+- Muted media files won't trigger Chrome media notifications.
+- Media notifications use the favicon if no artworks are defined.
```
I would rephrase:
`- If no artwork is define and there is a favicon at a desirable size, media notifications will use it.`

mounirlamouri replied:
You might also want to check your "artwork" vs "artworks".

xxyzzzq replied:
FYI, the full logic is:
If no artwork is defined, no suitable image is available, or image download failed, media notification will fallback to a favicon at a desirable size if available.

beaufortfrancois replied:
@xxyzzzq If image download failed, it doesn't always fallback to favicon right?

beaufortfrancois replied:
Oh you're right! Sorry for that ;)

beaufortfrancois replied:
Fixed.

xxyzzzq replied:
We fallback to favicon in this case. If not then it should be a bug :)

mounirlamouri commented on src/content/en/updates/2017/02/media-session.md:
```diff
@@ -0,0 +1,374 @@
+  [low-end devices], it is `256x256`.
+- Dismiss media notifications with `audio.src = ''`.
+- Hook up an `<audio>` element as the input source to the [Web Audio API] makes
+  it work with the Media Session API.
```
Should you have a short paragraph/section about audio focus? This might sound weird but if we were to explain that Media Session API only applies when Chrome will take audio focus could help some readers understand more deeply what's happening.

beaufortfrancois replied:
Do you mind sharing an abstract?

beaufortfrancois replied:
Done.

mounirlamouri commented on src/content/en/updates/2017/02/media-session.md:
```diff
@@ -0,0 +1,374 @@
+
+At the time of writing, Chrome for Android is the only platform that supports
+the Media Session API. More up-to-date information on browser implementation
+status can be found on [chromestatus.com].
```
Exact link to the entry maybe?

beaufortfrancois replied:
It is available at the very bottom of the page.

jakearchibald replied:
I agree this should be a link to the exact entry.

beaufortfrancois commented:
@xxyzzzq The hero image was still WIP as well. Next patch will include the "final" one.

beaufortfrancois commented:
I've just addressed @petele @xxyzzzq and @mounirlamouri feedback in https://github.com/google/WebFundamentals/pull/4057/commits/bd09396cac652ca30d30206b9b978719b43db384

beaufortfrancois commented:
@jakearchibald Do you mind having a look at this Media Session article? I'd love to get your feedback on the Service Worker / Cache part.

jakearchibald commented on src/content/en/updates/2017/02/media-session.md:
```diff
@@ -0,0 +1,384 @@
+With the brand new [Media Session API], you can now **customize media
+notifications** by providing metadata information such as the title, artist,
+album name, and artwork of the media (audio or video) your web app is playing
+in Chrome 57 (beta in February 2017, stable in March 2017). It also
```
This is a pretty long first sentence with a lot of commas and parenthesis. I'd lose "such as the title, artist,
 +album name, and artwork" and maybe "(audio or video)" too.

You can cover the details later in the article. In fact, the image below covers a lot of the details really well.

beaufortfrancois replied:
Done.

jakearchibald commented on src/content/en/updates/2017/02/media-session.md:
```diff
@@ -0,0 +1,398 @@
+## Gimme What I Want
+
+You already know all about the Media Session API and are simply coming back to
+copy and paste with no shame some boilerplate code? So here it is.
```
This is a great idea.

samdutton replied:
Yes!

jakearchibald commented on src/content/en/updates/2017/02/media-session.md:
```diff
@@ -0,0 +1,398 @@
+      <source src="audio.ogg" type="audio/ogg"/>
+    </audio>
+
+Note: I could have used a `<video>` tag as well there to showcase the Media
```
As well, or instead?

beaufortfrancois replied:
Done.

jakearchibald commented on src/content/en/updates/2017/02/media-session.md:
```diff
@@ -0,0 +1,398 @@
+Note: I could have used a `<video>` tag as well there to showcase the Media
+Session API.
+
+As you may know, `autoplay` is disabled on Chrome for Android which
```
It work for some types of media doesn't it? Dunno if that's worth clarifying.

beaufortfrancois replied:
I've added `for audio elements`. Thanks!

jakearchibald commented on src/content/en/updates/2017/02/media-session.md:
```diff
@@ -0,0 +1,398 @@
+Session API.
+
+As you may know, `autoplay` is disabled on Chrome for Android which
+means we have to use the `play()` method of the audio element there. This
```
"there" where?

beaufortfrancois replied:
Nowhere ;)

jakearchibald commented on src/content/en/updates/2017/02/media-session.md:
```diff
@@ -0,0 +1,398 @@
+  .catch(error => { console.log(error) });
+});
+
+</pre>
```
Is this section necessary? The audio element has `controls`, could we just rely on that? Then we can skip all the user interaction and "pointerup" stuff.

That stuff's important, but feels like it could be a different article.

In fact, does the code in this article fail if the user interacts with the audio element controls rather than `playButton`?

beaufortfrancois replied:
The play/load trick is not documented anywhere at all for the moment. This has beaten me when playing with 2 `<audio>` elements for instance.
I think developers playing with the Media Session API will be happy to learn these kinds of quirks.

beaufortfrancois replied:
I've added a note.

jakearchibald commented on src/content/en/updates/2017/02/media-session.md:
```diff
@@ -0,0 +1,398 @@
+### Customize the notification
+
+When your web app is playing audio, you can already see a media notification
+sitting in the notification tray. On Android, Chrome does its best there to show
```
I'd ditch the "there"

beaufortfrancois replied:
Done.

jakearchibald commented on src/content/en/updates/2017/02/media-session.md:
```diff
@@ -0,0 +1,398 @@
+
+<figure>
+  <img src="/web/updates/images/2017/02/tldr.png"
+    alt="Media Session TL;DR;"/>
```
There's a creative commons picture of Astley on wikipedia https://en.wikipedia.org/wiki/File:Rick_Astley_Tivoli_Gardens.jpg

beaufortfrancois replied:
I've added it.

jakearchibald commented on src/content/en/updates/2017/02/media-session.md:
```diff
@@ -0,0 +1,398 @@
+session metadata such as the title, artist, album name, and artwork with the
+Media Session API.
+
+    if ('mediaSession' in navigator) {
```
It's kinda covered in earlier code, but is it worth mentioning when this code should execute.

Maybe just a comment like:

```js
// When the audio starts playing…
```

beaufortfrancois replied:
Good idea!

jakearchibald commented on src/content/en/updates/2017/02/media-session.md:
```diff
@@ -0,0 +1,398 @@
+    let playlist = ['audio1.mp3', 'audio2.mp3', 'audio3.mp3'];
+    let index = 0;
+    
+    navigator.mediaSession.setActionHandler('previoustrack', function() {
```
Any idea why this is `setActionHandler` rather than `addEventListener`?

What if the user starts playing something without a playlist? Do I need to remove this handler somehow?

If I set the metadata to something else, do the handlers persist?

beaufortfrancois replied:
See https://github.com/WICG/mediasession/issues/148#issuecomment-267041447 for the all story about why not simply using `addEventListener`.

I've added a section to answer your questions.

jakearchibald commented on src/content/en/updates/2017/02/media-session.md:
```diff
@@ -0,0 +1,398 @@
+Forward" media notification icons if you want to control the amount of time
+skipped.
+
+    let skipTime= 10; /* Time to skip in seconds */
```
nit: space before `=`

jakearchibald replied:
Also, probably only need `//` rather than `/* … */`

beaufortfrancois replied:
Done. Done.

jakearchibald commented on src/content/en/updates/2017/02/media-session.md:
```diff
@@ -0,0 +1,398 @@
+      playAudio();
+    });
+
+The Media Session API allows you as well to show "Seek Backward" and "Seek
```
nit: Remove "as well"

It's kinda suggested here, but is it worth being explicit that these buttons won't be shown unless you add a handler?

beaufortfrancois replied:
Done.

jakearchibald commented on src/content/en/updates/2017/02/media-session.md:
```diff
@@ -0,0 +1,398 @@
+    .then(cache => cache.matchAll())
+    .then(responses => Promise.all(responses.map(r => r.blob())))
+    .then(blobs => blobs.map(blob => blob.size).reduce((acc, cur) => acc + cur, 0))
+    .then(cacheSize => { console.log('Artworks cache is ' + cacheSize + ' bytes.'); })
```
This is going to bring all the artworks into memory. Alternatively you could add up the `content-length` headers, and make estimates where this header is missing.

beaufortfrancois replied:
Here's the code below I've used to add up content-length headers. It gives me different results than using `blob.size()`...
 
```js
caches.open('web-bluetooth')
.then(cache => cache.matchAll())
.then(responses => {
  let cacheSize = 0;
  let queue = Promise.resolve();
  responses.forEach(response => {
    let responseSize = Number(response.headers.get('content-length'));
    if (responseSize) {
      queue = queue.then(_ => cacheSize += responseSize);
    } else {
      queue = queue.then(_ => response.blob())
          .then(blob => cacheSize += blob.size);
    }
  });
  return queue.then(_ => {
    console.log('Artworks cache is ' + cacheSize + ' Bytes.'); 
  });
})
.catch(error => { console.log(error); });
```

beaufortfrancois replied:
@jakearchibald What do you think of using this code?

```javascript
caches.open('artworks-cache-v1')
.then(cache => cache.matchAll())
.then(responses => {
  let cacheSize = 0;
  let queue = Promise.resolve();
  responses.forEach(response => {
    queue = queue.then(_ => response.blob())
          .then(blob => { cacheSize += blob.size; blob.close(); });
  });
  return queue.then(_ => {
    console.log('Artworks cache is ' + cacheSize + ' Bytes.'); 
  });
})
.catch(error => { console.log(error); });
```

beaufortfrancois replied:
https://github.com/google/WebFundamentals/pull/4057/files/dce240d58480a124e1817e80ecefdaa3612f68b9#r97076285 is "correct". I'll use this!

beaufortfrancois replied:
And here's https://github.com/w3c/ServiceWorker/issues/1059

jakearchibald commented on src/content/en/updates/2017/02/media-session.md:
```diff
@@ -0,0 +1,398 @@
+
+## Samples & Demos
+
+Check out our official [Media Session Chrome samples].
```
Broken link

beaufortfrancois replied:
The media session sample has not been reviewed yet.

jakearchibald commented on src/content/en/updates/2017/02/media-session.md:
```diff
@@ -0,0 +1,398 @@
+album name, and artwork of the media (audio or video) your web app is playing
+in Chrome 57 (beta in February 2017, stable in March 2017). It also
+allows you to **handle media related events** such as seeking or track
+changing which may come from notifications or media keys.
```
Maybe link to a demo here? As in, show it working, then go on to explain how it works.

beaufortfrancois replied:
Done.

jakearchibald commented:
Does the developer have to do anything to release the session once playback is done?

jakearchibald commented:
Really enjoyed this article!

jakearchibald commented:
However, dare I say it… is the service worker stuff a little off-topic? (didn't want to say this, because, of course I want everything to be about service worker)

beaufortfrancois commented:
Thanks for reviewing @jakearchibald.
I've addressed all your feedback except for the blob's one yet as I'd like to find a way to not rely on `content-length`.

Re: Service Worker, I think the offline story needs to be told "there". It makes sense especially when you play with the Media Session API on a lie-fi connection.

beaufortfrancois commented:
@petele Is there a staging server for PRs where we can tell people to look at?

beaufortfrancois commented:
@jpmedley It is not yet finalized but I think it is in a "good shape" for you to have a look. Thank you.

petele commented:
@beaufortfrancois - sadly no staging server for folks to look at right now. :(

samdutton commented on src/content/en/updates/2017/02/media-session.md:
```diff
@@ -0,0 +1,445 @@
+{% include "web/_shared/contributors/beaufortfrancois.html" %}
+
+With the brand new [Media Session API], you can now **customize media
+notifications** by providing metadata information of the media your web app is
```
Nit: I think this should be 'for the media' not 'of the media'.

beaufortfrancois replied:
Done.

samdutton commented on src/content/en/updates/2017/02/media-session.md:
```diff
@@ -0,0 +1,445 @@
+
+With the brand new [Media Session API], you can now **customize media
+notifications** by providing metadata information of the media your web app is
+playing in Chrome 57 (beta in February 2017, stable in March 2017). It also
```
I think you should probably add a full stop after 'playing' and then start a new sentence with '`MediaSession` is supported in Chrome 57 ...'

beaufortfrancois replied:
I've added a full stop and moved the "supported" sentence. WDYT?

samdutton commented on src/content/en/updates/2017/02/media-session.md:
```diff
@@ -0,0 +1,445 @@
+      <source src="audio.ogg" type="audio/ogg"/>
+    </audio>
+
+Note: I could have used a `<video>` tag instead to showcase the Media Session
```
Nit: `the MediaSession API.'

beaufortfrancois replied:
It is actually the "Media Session API", not "MediaSession API" according to https://wicg.github.io/mediasession/

samdutton commented on src/content/en/updates/2017/02/media-session.md:
```diff
@@ -0,0 +1,445 @@
+
+### Let's play 🎷
+
+Add a simple `<audio>` tag to your web page and assign several media sources so
```
Nit: should probably say '`<audio>` element' and '`<video>` element'.

beaufortfrancois replied:
Thanks! Done.

samdutton commented on src/content/en/updates/2017/02/media-session.md:
```diff
@@ -0,0 +1,445 @@
+    });
+
+Note: If the `<audio>` element has the `controls` attribute, you can simply set
+up the Media Session in the audio `play` event listener instead which occurs
```
MediaSession?

samdutton replied:
Might be worth explaining **why** you might do it that way.

beaufortfrancois replied:
Nope ;)

beaufortfrancois replied:
*I* wouldn't actually do it that way.
Why would you do?

samdutton replied:
I don't feel strongly either way, but it's not really clear from the text what the benefits are of one way or the other.  No big deal though...

jpmedley replied:
The rules are these:
* When referring to the API proper, capitalize. Media Session API
* When referring to a generic media session, do not capitalize. media session
* When referring to an instance of the `MediaSession` interface, like this: `mediaSession` (or something similar)
* When referring to the `MediaSession` interface itself, like this: `MediaSession`

Please verify throughout.

samdutton commented on src/content/en/updates/2017/02/media-session.md:
```diff
@@ -0,0 +1,445 @@
+If you don't want to play audio right after the first interaction, I recommend
+you use the `load()` method of the audio element. This is one way for the
+browser to keep track of whether the user interacted with the element. Note
+that it may also help the playback to be smoother because the content will be
```
Nit: maybe 'help playback' and 'will already be loaded.'

beaufortfrancois replied:
Done.

samdutton commented on src/content/en/updates/2017/02/media-session.md:
```diff
@@ -0,0 +1,445 @@
+
+When your web app is playing audio, you can already see a media notification
+sitting in the notification tray. On Android, Chrome does its best to show
+appropriate information by using the document's title and the largest favicon
```
Nit: it's not a favicon, is it, just an 'icon image'?

beaufortfrancois replied:
You're right! Thanks

samdutton commented on src/content/en/updates/2017/02/media-session.md:
```diff
@@ -0,0 +1,445 @@
+<div class="attempt-left">
+  <figure>
+    <img src="/web/updates/images/2017/02/without-media-session.png" alt="Without Media Session">
+    <figcaption>With no Media Session</figcaption>
```
Maybe MediaSession rather than Media Session (throughout)?

beaufortfrancois replied:
Still nope ;)

samdutton replied:
np :)

samdutton commented on src/content/en/updates/2017/02/media-session.md:
```diff
@@ -0,0 +1,445 @@
+</div>
+<div class="clearfix"></div>
+
+#### Set Metadata
```
Nit: Set metadata

beaufortfrancois replied:
Thanks!

samdutton commented on src/content/en/updates/2017/02/media-session.md:
```diff
@@ -0,0 +1,445 @@
+
+#### Previous Track / Next Track
+
+If your web app provides a playlist for instance, you may want to allow the
```
Maybe delete 'for instance'.

beaufortfrancois replied:
Deleted ;)

samdutton commented on src/content/en/updates/2017/02/media-session.md:
```diff
@@ -0,0 +1,445 @@
+
+
+Note that media action handlers will persist. This is very similar to the event
+listener pattern except that handling an event means that the browser steps
```
'stops doing any default ...'

beaufortfrancois replied:
Done.

samdutton commented on src/content/en/updates/2017/02/media-session.md:
```diff
@@ -0,0 +1,445 @@
+      // Do something more than just pausing current audio...
+    });
+
+Note: As the browser may consider the web app is not be playing when media
```
'The browser may consider that the web app is not playing media when files are seeking or loading. You can override ...'

beaufortfrancois replied:
Thanks.

samdutton commented on src/content/en/updates/2017/02/media-session.md:
```diff
@@ -0,0 +1,445 @@
+Note: As the browser may consider the web app is not be playing when media
+files are seeking or loading, you can override this behaviour by setting
+`navigator.mediaSession.playbackState` to `"playing"` or `"paused"`.  This
+comes handy when you want to make sure your web app UI stays in sync with the
```
'comes in handy'

beaufortfrancois replied:
Done.

samdutton commented on src/content/en/updates/2017/02/media-session.md:
```diff
@@ -0,0 +1,445 @@
+### The Service Worker caching strategy
+
+Regarding media files, I would recommend a simple "[Cache, falling back to
+network]" strategy as illustrated by Jake Archibald.
```
Maybe 'as described by Jake Archibald', with a link.

beaufortfrancois replied:
`[Cache, falling back to network]` is already a link.

samdutton replied:
Ah — thanks.

samdutton commented on src/content/en/updates/2017/02/media-session.md:
```diff
@@ -0,0 +1,445 @@
+Regarding media files, I would recommend a simple "[Cache, falling back to
+network]" strategy as illustrated by Jake Archibald.
+
+For artworks though, I'd be a little bit more specific and choose the approach
```
Nit: I think 'artwork' is better — 'artworks' sounds like works of art :).

beaufortfrancois replied:
Oops ;)
Done.

samdutton commented on src/content/en/updates/2017/02/media-session.md:
```diff
@@ -0,0 +1,445 @@
+
+### Let user control cache
+
+As user will consume content from your web app, media files and artworks may
```
'As the user consumes content from your web app, media files and artwork may take a lot of space on their device.'

beaufortfrancois replied:
Thanks!

samdutton commented on src/content/en/updates/2017/02/media-session.md:
```diff
@@ -0,0 +1,445 @@
+## Implementation nits
+
+- Chrome for Android requests "full" audio focus to show media notifications
+  only when media files duration is [at least 5 seconds].
```
'media file duration'

beaufortfrancois replied:
Oops.

samdutton commented on src/content/en/updates/2017/02/media-session.md:
```diff
@@ -0,0 +1,445 @@
+
+- Chrome for Android requests "full" audio focus to show media notifications
+  only when media files duration is [at least 5 seconds].
+- If no artwork is defined and there is a favicon at a desirable size, media
```
Probably should be 'icon' not 'favicon'.

samdutton replied:
Do you really mean favicon (i.e. like simpl.info/favicon.ico)?

beaufortfrancois replied:
`icon image` will be ;)

beaufortfrancois commented:
@samdutton I've addressed all your feedback. Thanks again!

jpmedley commented:
@beaufortfrancois Just after you asked me to edit was when Sam put in all of his comments. I want to double-check. Are you expecting feedback from ANYONE else?

@jakearchibald I'll need you to review and approve  your requested changes as well.

beaufortfrancois commented:
As discussed offline, @jakearchibald is happy for it to launch as it.
Therefore, I'd love your final review @jpmedley.

jakearchibald commented:
Yep! The article was great before I reviewed it. Any fixes are a bonus 😄

petele commented:
@jakearchibald - can you please "approve" the changes in git so that I don't have to "dismiss" your review before I merge? :)

jpmedley commented on src/content/en/updates/2017/02/media-session.md:
```diff
@@ -0,0 +1,454 @@
+    alt="Media Session TL;DR;"/>
+</figure>
+
+## Gimme What I Want
```
All titles level 2 and lower should be sentence case. Only the article title should be title case. Please change throughout.

beaufortfrancois replied:
Done.

jpmedley commented on src/content/en/updates/2017/02/media-session.md:
```diff
@@ -0,0 +1,454 @@
+is why you need to update it to make sure you're always showing relevant
+information in the media notification.
+
+#### Previous Track / Next Track
```
Sentence case.

beaufortfrancois replied:
Done.

jpmedley commented on src/content/en/updates/2017/02/media-session.md:
```diff
@@ -0,0 +1,454 @@
+True but first and foremost, you want to make sure **all items in this
+checklist are checked**:
+
+1. All media and artwork files are served with the appropriate `Cache-Control`
```
Items that aren't required to be in a particular order should be bulleted, not numbered.

beaufortfrancois replied:
Done.

jpmedley commented on src/content/en/updates/2017/02/media-session.md:
```diff
@@ -0,0 +1,454 @@
+`Allow-Control-Allow-Origin: *` HTTP header. This will allow third-party web
+apps to fetch and consume HTTP responses from your web server.
+
+### The Service Worker caching strategy
```
The term service worker is lower case unless referring to the 'Service Worker API'. See my rules for media source, above. Please check throughout.

beaufortfrancois replied:
Done.

jpmedley commented on src/content/en/updates/2017/02/media-session.md:
```diff
@@ -0,0 +1,454 @@
+
+### The Service Worker caching strategy
+
+Regarding media files, I would recommend a simple "[Cache, falling back to
```
Delete the word 'would'.

beaufortfrancois replied:
Done.

jpmedley commented on src/content/en/updates/2017/02/media-session.md:
```diff
@@ -0,0 +1,454 @@
+    .then(cache => Promise.all(artworkFilesToDelete.map(artwork => cache.delete(artwork))))
+    .catch(error => { console.log(error); });
+
+## Implementation nits
```
I think 'Implementation notes' would be a better title.

beaufortfrancois replied:
Done.

jpmedley commented on src/content/en/updates/2017/02/media-session.md:
```diff
@@ -0,0 +1,454 @@
+
+- Chrome for Android requests "full" audio focus to show media notifications
+  only when media file duration is [at least 5 seconds].
+- If no artwork is defined and there is an icon image at a desirable size, media
```
Do  you mean appropriate size?

beaufortfrancois commented:
@jpmedley Thank you very much. I've addressed your valuable feedback at https://github.com/google/WebFundamentals/pull/4057/commits/460bfb3015cd34f0625191dd41bc0322348e0ab2

beaufortfrancois commented:
If @jpmedley is OK,  @petele please merge and push live.
